pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go

```diff
@@ -0,0 +1,361 @@
+	if err != nil {
+		return fmt.Errorf("couldn't create builds for architectures: %s: %w", strings.Join(r.architectures, ","), err)
+	}
+
+	if err := r.watchAndHandleBuildEvents(ctx, mabc, builds); err != nil {
+		return fmt.Errorf("could't watch builds: %w", err)
+	}
+
+	mutateFn := func(mabcToMutate *v1.MultiArchBuildConfig) { mabcToMutate.Status.State = v1.FailureState }
+	if checkAllBuildsSuccessful(builds) {
+		if err := pushOutputImageWithManifest(builds, targetImageRef); err != nil {
+			mutateFn = func(mabcToMutate *v1.MultiArchBuildConfig) {
+				mabcToMutate.Status.Conditions = append(mabcToMutate.Status.Conditions, metav1.Condition{
+					Type:               "PushManifestError",
+					Status:             metav1.ConditionFalse,
+					LastTransitionTime: metav1.Time{Time: time.Now()},
+					Reason:             "PushManifestError",
+					Message:            err.Error(),
+				})
+			}
+		} else {
+			mutateFn = func(mabcToMutate *v1.MultiArchBuildConfig) { mabcToMutate.Status.State = v1.SuccessState }
+		}
+	}
+	if err := r.updateMultiArchBuildConfig(ctx, ctrlruntimeclient.ObjectKey{Namespace: mabc.Namespace, Name: mabc.Name}, mutateFn); err != nil {
+		return fmt.Errorf("failed to update the MultiArchBuildConfig %s/%s: %w", mabc.Namespace, mabc.Name, err)
+	}
+
+	if err := r.pruneBuilds(ctx, builds); err != nil {
+		return fmt.Errorf("failed to prune builds: %w", err)
+	}
+
+	r.logger.WithField("mabc_name", mabc.Name).Info("All builds have been handled")
+	return nil
+}
+
+func (r *reconciler) watchAndHandleBuildEvents(ctx context.Context, mabc *v1.MultiArchBuildConfig, builds map[string]*buildv1.Build) error {
+	r.logger.Info("Watching builds events...")
+	w, err := r.buildClient.Builds(mabc.Namespace).Watch(ctx, metav1.ListOptions{LabelSelector: labels.Set{v1.MultiArchBuildConfigNameLabel: mabc.Name}.String()})
+	if err != nil {
+		return err
+	}
+	defer w.Stop()
+
+	for {
+		event, ok := <-w.ResultChan()
+		if !ok {
+			return fmt.Errorf("watch channel closed")
+		}
+		r.logger.Debugf("Event received... %s", event.Type)
+
+		build := event.Object.(*buildv1.Build)
+		switch event.Type {
+		case watch.Deleted:
+			delete(builds, build.Name)
+		case watch.Added, watch.Modified:
+			r.logger.WithField("build_name", build.Name).WithField("build_phase", build.Status.Phase).Debug("Storing build")
+			builds[fmt.Sprintf("%s/%s", build.Namespace, build.Name)] = build
+
+			if err := r.updateMultiArchBuildConfig(ctx, ctrlruntimeclient.ObjectKey{Namespace: mabc.Namespace, Name: mabc.Name},
+				func(mabcToMutate *v1.MultiArchBuildConfig) {
+					mabcToMutate.Status.Conditions = append(mabcToMutate.Status.Conditions, constructCondition(build))
+				}); err != nil {
+				return err
+			}
+		}
+
+		if checkAllBuildsFinished(builds) {
+			r.logger.Info("All builds have finished...")
+			return nil
+		}
+	}
+}
+
+func (r *reconciler) createBuildsForArchitectures(ctx context.Context, mabc *v1.MultiArchBuildConfig) (map[string]*buildv1.Build, error) {
```

Comment:
As a consequence, this one is also not covered.

Reply:
I will add tests for this as well.